libraries/chain/block_state.cpp

```diff
@@ -9,20 +9,66 @@
+                             const block_signing_authority& valid_block_signing_authority)
+   {
+      auto num_keys_in_authority = std::visit([](const auto& a) { return a.keys.size(); },
+                                              valid_block_signing_authority);
+      EOS_ASSERT(1 + additional_signatures.size() <= num_keys_in_authority, wrong_signing_key,
+                 "number of block signatures (${num_block_signatures}) exceeds number of keys (${num_keys}) in block"
+                 " signing authority: ${authority}",
+                 ("num_block_signatures", 1 + additional_signatures.size())("num_keys", num_keys_in_authority)
+                 ("authority", valid_block_signing_authority));
+
+      std::set<public_key_type> keys;
+      keys.emplace(fc::crypto::public_key(block->producer_signature, block_id, true));
+
+      for (const auto& s : additional_signatures) {
+         auto res = keys.emplace(s, block_id, true);
+         EOS_ASSERT(res.second, wrong_signing_key, "block signed by same key twice: ${key}", ("key", *res.first));
+      }
+
+      bool is_satisfied = false;
+      size_t relevant_sig_count = 0;
+
+      std::tie(is_satisfied, relevant_sig_count) = producer_authority::keys_satisfy_and_relevant(
+         keys, valid_block_signing_authority);
+
+      EOS_ASSERT(relevant_sig_count == keys.size(), wrong_signing_key,
+                 "block signed by unexpected key: ${signing_keys}, expected: ${authority}. ${c} != ${s}",
+                 ("signing_keys", keys)("authority", valid_block_signing_authority)("c", relevant_sig_count)("s", keys. size()));
+
+      EOS_ASSERT(is_satisfied, wrong_signing_key,
+                 "block signatures ${signing_keys} do not satisfy the block signing authority: ${authority}",
+                 ("signing_keys", keys)("authority", valid_block_signing_authority));
+   }
+
+   // EOS_ASSERTs if signature does not validate
+   inline bool verify_block_sig(const block_header_state& prev, const signed_block_ptr& block, bool skip_validate_signee) {
+      if (!skip_validate_signee) {
+         auto sigs = detail::extract_additional_signatures(block);
+         const auto& valid_block_signing_authority = prev.get_producer_for_block_at(block->timestamp).authority;
+         verify_signee(block, block->calculate_id(), sigs, valid_block_signing_authority);
+      }
+      return true;
+   };
+
+} // namespace detail
+
+using namespace eosio::chain::detail;
+
+// ASSUMPTION FROM controller_impl::apply_block = all untrusted blocks will have their signatures pre-validated here
 block_state::block_state(const block_header_state& prev, signed_block_ptr b, const protocol_feature_set& pfs,
                          const validator_t& validator, bool skip_validate_signee)
-   : block_header_state(prev.next(*b, validator))
+   : block_header_state(verify_block_sig(prev, b, skip_validate_signee) ? prev.next(*b, validator) : block_header_state{})
```

Comment:
`verify_block_sig` always returns true, so the ternary operator is weird. I think the following is clearer maybe:
```suggestion
   : block_header_state(verify_block_sig(prev, b, skip_validate_signee), prev.next(*b, validator))
```

Reply:
I tried that first, but compiler doesn't like it:
```
/home/heifner/ext/leap/libraries/chain/block_state.cpp:66:25: error: no viable conversion from 'bool' to 'block_id_type' (aka 'fc::sha256')
   : block_header_state(verify_block_sig(prev, b, skip_validate_signee), prev.next(*b, validator))
```
However, an extra set of `()` seems to do the trick.